File: .github/ai-team/scripts/cost-reporter.js
#!/usr/bin/env node

/**
 * Cost Reporter for GitHub Actions AI Team
 * Generates detailed cost reports and insights
 */

const fs = require('fs');
const path = require('path');

class CostReporter {
  constructor() {
    this.costTrackerPath = process.env.GITHUB_WORKSPACE 
      ? path.join(process.env.GITHUB_WORKSPACE, '.github/ai-team/cost-tracker.json')
      : path.join(__dirname, '../cost-tracker.json');
      
    this.configPath = path.join(__dirname, '../configs/main-config.yml');
  }
  
  generateReport(period = 'daily') {
    const costData = this.loadCostData();
    const config = this.loadConfig();
    
    const now = new Date();
    const report = {
      generated: now.toISOString(),
      period: period,
      budget: config.budget,
      data: {}
    };
    
    switch (period) {
      case 'daily':
        report.data = this.generateDailyReport(costData, now);
        break;
      case 'weekly':
        report.data = this.generateWeeklyReport(costData, now);
        break;
      case 'monthly':
        report.data = this.generateMonthlyReport(costData, now);
        break;
      default:
        throw new Error(`Unknown period: ${period}`);
    }
    
    return report;
  }
  
  generateDailyReport(costData, now) {
    const today = now.toISOString().slice(0, 10);
    const dailyData = costData.daily[today] || { total: 0, bots: {} };
    
    return {
      date: today,
      total: dailyData.total,
      bots: dailyData.bots,
      topSpenders: this.getTopSpenders(dailyData.bots, 5),
      alerts: this.checkDailyAlerts(dailyData)
    };
  }
  
  generateWeeklyReport(costData, now) {
    const weekData = { total: 0, bots: {}, days: {} };
    
    // Get last 7 days
    for (let i = 0; i < 7; i++) {
      const date = new Date(now);
      date.setDate(date.getDate() - i);
      const dateStr = date.toISOString().slice(0, 10);
      
      if (costData.daily[dateStr]) {
        weekData.total += costData.daily[dateStr].total;
        weekData.days[dateStr] = costData.daily[dateStr].total;
        
        // Aggregate bot costs
        for (const [bot, cost] of Object.entries(costData.daily[dateStr].bots)) {
          weekData.bots[bot] = (weekData.bots[bot] || 0) + cost;
        }
      }
    }
    
    return {
      startDate: new Date(now.getTime() - 6 * 24 * 60 * 60 * 1000).toISOString().slice(0, 10),
      endDate: now.toISOString().slice(0, 10),
      total: weekData.total,
      dailyAverage: weekData.total / 7,
      bots: weekData.bots,
      topSpenders: this.getTopSpenders(weekData.bots, 5),
      dailyBreakdown: weekData.days,
      trend: this.calculateTrend(weekData.days)
    };
  }
  
  generateMonthlyReport(costData, now) {
    const currentMonth = now.toISOString().slice(0, 7);
    const monthData = costData.monthly[currentMonth] || { total: 0, bots: {} };
    
    // Calculate projection
    const dayOfMonth = now.getDate();
    const daysInMonth = new Date(now.getFullYear(), now.getMonth() + 1, 0).getDate();
    const projectedTotal = (monthData.total / dayOfMonth) * daysInMonth;
    
    return {
      month: currentMonth,
      total: monthData.total,
      projected: projectedTotal,
      remaining: 100 - monthData.total, // Assuming $100 budget
      bots: monthData.bots,
      topSpenders: this.getTopSpenders(monthData.bots, 10),
      budgetStatus: this.getBudgetStatus(monthData.total, projectedTotal),
      recommendations: this.getRecommendations(monthData, projectedTotal)
    };
  }
  
  getTopSpenders(bots, limit) {
    const total = Object.values(bots).reduce((a, b) => a + b, 0);
    
    // Guard against division by zero when bots object is empty
    if (total === 0) {
      return [];
    }
    
    return Object.entries(bots)
      .sort(([, a], [, b]) => b - a)
      .slice(0, limit)
      .map(([bot, cost]) => ({ bot, cost, percentage: (cost / total * 100).toFixed(1) }));
  }
  
  checkDailyAlerts(dailyData) {
    const alerts = [];
    
    if (dailyData.total > 5) {
      alerts.push({
        level: 'warning',
        message: `Daily spend ($${dailyData.total.toFixed(2)}) exceeds $5 threshold`
      });
    }
    
    // Check for unusual bot activity
    for (const [bot, cost] of Object.entries(dailyData.bots)) {
      if (cost > 2) {
        alerts.push({
          level: 'info',
          message: `${bot} spent $${cost.toFixed(2)} today (above $2 threshold)`
        });
      }
    }
    
    return alerts;
  }
  
  calculateTrend(dailyData) {
    const values = Object.values(dailyData);
    if (values.length < 2) return 'insufficient_data';
    
    const firstHalf = values.slice(0, Math.floor(values.length / 2));
    const secondHalf = values.slice(Math.floor(values.length / 2));
    
    const firstAvg = firstHalf.reduce((a, b) => a + b, 0) / firstHalf.length;
    const secondAvg = secondHalf.reduce((a, b) => a + b, 0) / secondHalf.length;
    
    const change = ((secondAvg - firstAvg) / firstAvg) * 100;
    
    if (change > 20) return 'increasing_rapidly';
    if (change > 5) return 'increasing';
    if (change < -20) return 'decreasing_rapidly';
    if (change < -5) return 'decreasing';
    return 'stable';
  }
  
  getBudgetStatus(current, projected) {
    const budget = 100; // $100 monthly budget
    const percentUsed = (current / budget) * 100;
    const projectedPercent = (projected / budget) * 100;
    
    return {
      percentUsed: percentUsed.toFixed(1),
      projectedPercent: projectedPercent.toFixed(1),
      status: projectedPercent > 100 ? 'over_budget' : 
              projectedPercent > 90 ? 'near_limit' :
              projectedPercent > 75 ? 'on_track' : 'under_budget',
      daysRemaining: new Date(new Date().getFullYear(), new Date().getMonth() + 1, 0).getDate() - new Date().getDate()
    };
  }
  
  getRecommendations(monthData, projectedTotal) {
    const recommendations = [];
    const budget = 100;
    
    if (projectedTotal > budget * 1.1) {
      recommendations.push({
        priority: 'high',
        action: 'Reduce usage of expensive models',
        impact: 'Could save $' + ((projectedTotal - budget) * 0.5).toFixed(2)
      });
    }
    
    // Find expensive bots
    const expensiveBots = Object.entries(monthData.bots)
      .filter(([, cost]) => cost > budget * 0.2)
      .map(([bot]) => bot);
      
    if (expensiveBots.length > 0) {
      recommendations.push({
        priority: 'medium',
        action: `Optimize ${expensiveBots.join(', ')} - using >20% of budget each`,
        impact: 'Better cost distribution'
      });
    }
    
    if (projectedTotal < budget * 0.5) {
      recommendations.push({
        priority: 'low',
        action: 'Consider using more capable models for better results',
        impact: 'Improved quality without exceeding budget'
      });
    }
    
    return recommendations;
  }
  
  formatMarkdownReport(report) {
    let markdown = `# 💰 AI Team Cost Report\n\n`;
    markdown += `**Generated**: ${new Date(report.generated).toLocaleString()}\n`;
    markdown += `**Period**: ${report.period}\n`;
    markdown += `**Budget**: $${report.budget.monthly_limit}/month\n\n`;
    
    const data = report.data;
    
    switch (report.period) {
      case 'daily':
        markdown += `## 📊 Daily Report (${data.date})\n\n`;
        markdown += `**Total Spend**: $${data.total.toFixed(2)}\n\n`;
        
        if (data.topSpenders.length > 0) {
          markdown += `### 🏆 Top Spenders\n\n`;
          markdown += `| Bot | Cost | % of Total |\n`;
          markdown += `|-----|------|------------|\n`;
          data.topSpenders.forEach(s => {
            markdown += `| ${s.bot} | $${s.cost.toFixed(2)} | ${s.percentage}% |\n`;
          });
          markdown += `\n`;
        }
        
        if (data.alerts.length > 0) {
          markdown += `### ⚠️ Alerts\n\n`;
          data.alerts.forEach(alert => {
            markdown += `- **${alert.level.toUpperCase()}**: ${alert.message}\n`;
          });
        }
        break;
        
      case 'weekly':
        markdown += `## 📊 Weekly Report (${data.startDate} to ${data.endDate})\n\n`;
        markdown += `**Total Spend**: $${data.total.toFixed(2)}\n`;
        markdown += `**Daily Average**: $${data.dailyAverage.toFixed(2)}\n`;
        markdown += `**Trend**: ${data.trend.replace(/_/g, ' ')}\n\n`;
        
        markdown += `### 📈 Daily Breakdown\n\n`;
        markdown += `\`\`\`\n`;
        Object.entries(data.dailyBreakdown).forEach(([date, cost]) => {
          const bar = '█'.repeat(Math.round(cost * 2));
          markdown += `${date}: ${bar} $${cost.toFixed(2)}\n`;
        });
        markdown += `\`\`\`\n\n`;
        
        if (data.topSpenders.length > 0) {
          markdown += `### 🏆 Top Spenders\n\n`;
          markdown += `| Bot | Cost | % of Total |\n`;
          markdown += `|-----|------|------------|\n`;
          data.topSpenders.forEach(s => {
            markdown += `| ${s.bot} | $${s.cost.toFixed(2)} | ${s.percentage}% |\n`;
          });
        }
        break;
        
      case 'monthly':
        markdown += `## 📊 Monthly Report (${data.month})\n\n`;
        markdown += `**Current Spend**: $${data.total.toFixed(2)} (${data.budgetStatus.percentUsed}% of budget)\n`;
        markdown += `**Projected Total**: $${data.projected.toFixed(2)} (${data.budgetStatus.projectedPercent}%)\n`;
        markdown += `**Budget Status**: ${data.budgetStatus.status.replace(/_/g, ' ').toUpperCase()}\n`;
        markdown += `**Days Remaining**: ${data.budgetStatus.daysRemaining}\n\n`;
        
        // Visual budget bar
        const used = Math.round(parseFloat(data.budgetStatus.percentUsed) / 5);
        const projected = Math.round(parseFloat(data.budgetStatus.projectedPercent) / 5);
        markdown += `### 💵 Budget Visualization\n\n`;
        markdown += `Current:   ${'█'.repeat(used)}${'░'.repeat(20 - used)} ${data.budgetStatus.percentUsed}%\n`;
        markdown += `Projected: ${'█'.repeat(projected)}${'░'.repeat(20 - projected)} ${data.budgetStatus.projectedPercent}%\n\n`;
        
        if (data.topSpenders.length > 0) {
          markdown += `### 🏆 Top Spending Bots\n\n`;
          markdown += `| Bot | Cost | % of Total |\n`;
          markdown += `|-----|------|------------|\n`;
          data.topSpenders.forEach(s => {
            markdown += `| ${s.bot} | $${s.cost.toFixed(2)} | ${s.percentage}% |\n`;
          });
          markdown += `\n`;
        }
        
        if (data.recommendations.length > 0) {
          markdown += `### 💡 Recommendations\n\n`;
          data.recommendations.forEach(rec => {
            markdown += `**${rec.priority.toUpperCase()}**: ${rec.action}\n`;
            markdown += `- Impact: ${rec.impact}\n\n`;
          });
        }
        break;
    }
    
    markdown += `\n---\n`;
    markdown += `*Cost tracking by CFO Bot • [View Full Dashboard](https://github.com/${process.env.GITHUB_REPOSITORY}/actions/workflows/cost-tracker.yml)*`;
    
    return markdown;
  }
  
  loadCostData() {
    try {
      if (fs.existsSync(this.costTrackerPath)) {
        return JSON.parse(fs.readFileSync(this.costTrackerPath, 'utf8'));
      }
    } catch (e) {
      console.error('Error loading cost data:', e.message);
    }
    return { daily: {}, monthly: {}, bots: {} };
  }
  
  loadConfig() {
    const yaml = require('js-yaml');
    const configContent = fs.readFileSync(this.configPath, 'utf8');
    return yaml.load(configContent);
  }
}

// Export for use in workflows
module.exports = CostReporter;

// CLI interface
if (require.main === module) {
  const reporter = new CostReporter();
  const period = process.argv[2] || 'daily';
  
  try {
    const report = reporter.generateReport(period);
    const markdown = reporter.formatMarkdownReport(report);
    console.log(markdown);
  } catch (error) {
    console.error('Error generating report:', error.message);
    process.exit(1);
  }
}

File: .github/ai-team/scripts/cost_reporter.py
#!/usr/bin/env python3
"""
Cost Reporter for GitHub Actions AI Team
Generates detailed cost reports and insights
"""

import json
import os
import sys
from datetime import datetime, timedelta
from pathlib import Path
from typing import Dict, Any, List, Optional
import yaml


class CostReporter:
    """Generates cost reports for the AI team."""
    
    def __init__(self):
        """Initialize the cost reporter."""
        workspace = os.environ.get('GITHUB_WORKSPACE')
        if workspace:
            self.cost_tracker_path = Path(workspace) / '.github' / 'ai-team' / 'cost-tracker.json'
        else:
            script_dir = Path(__file__).parent
            self.cost_tracker_path = script_dir.parent / 'cost-tracker.json'
        
        self.config_path = Path(__file__).parent.parent / 'configs' / 'main-config.yml'
    
    def load_cost_data(self) -> Dict[str, Any]:
        """Load cost tracking data from JSON file."""
        try:
            if self.cost_tracker_path.exists():
                with open(self.cost_tracker_path, 'r', encoding='utf-8') as f:
                    return json.load(f)
        except Exception as e:
            print(f'Error loading cost data: {e}', file=sys.stderr)
        return {'daily': {}, 'monthly': {}, 'bots': {}}
    
    def load_config(self) -> Dict[str, Any]:
        """Load configuration from YAML file."""
        with open(self.config_path, 'r', encoding='utf-8') as f:
            return yaml.safe_load(f)
    
    def generate_report(self, period: str = 'daily') -> Dict[str, Any]:
        """Generate cost report for the specified period."""
        cost_data = self.load_cost_data()
        config = self.load_config()
        
        now = datetime.now()
        report = {
            'generated': now.isoformat(),
            'period': period,
            'budget': config['budget'],
            'data': {}
        }
        
        if period == 'daily':
            report['data'] = self.generate_daily_report(cost_data, now)
        elif period == 'weekly':
            report['data'] = self.generate_weekly_report(cost_data, now)
        elif period == 'monthly':
            report['data'] = self.generate_monthly_report(cost_data, now)
        else:
            raise ValueError(f'Unknown period: {period}')
        
        return report
    
    def generate_daily_report(self, cost_data: Dict[str, Any], now: datetime) -> Dict[str, Any]:
        """Generate daily cost report."""
        today = now.strftime('%Y-%m-%d')
        daily_data = cost_data.get('daily', {}).get(today, {'total': 0.0, 'bots': {}})
        
        return {
            'date': today,
            'total': daily_data.get('total', 0.0),
            'bots': daily_data.get('bots', {}),
            'top_spenders': self.get_top_spenders(daily_data.get('bots', {}), 5),
            'alerts': self.check_daily_alerts(daily_data)
        }
    
    def generate_weekly_report(self, cost_data: Dict[str, Any], now: datetime) -> Dict[str, Any]:
        """Generate weekly cost report."""
        week_data = {'total': 0.0, 'bots': {}, 'days': {}}
        
        # Get last 7 days
        for i in range(7):
            date = now - timedelta(days=i)
            date_str = date.strftime('%Y-%m-%d')
            
            if date_str in cost_data.get('daily', {}):
                day_data = cost_data['daily'][date_str]
                week_data['total'] += day_data.get('total', 0.0)
                week_data['days'][date_str] = day_data.get('total', 0.0)
                
                # Aggregate bot costs
                for bot, cost in day_data.get('bots', {}).items():
                    week_data['bots'][bot] = week_data['bots'].get(bot, 0.0) + cost
        
        start_date = (now - timedelta(days=6)).strftime('%Y-%m-%d')
        end_date = now.strftime('%Y-%m-%d')
        
        return {
            'start_date': start_date,
            'end_date': end_date,
            'total': week_data['total'],
            'daily_average': week_data['total'] / 7,
            'bots': week_data['bots'],
            'top_spenders': self.get_top_spenders(week_data['bots'], 5),
            'daily_breakdown': week_data['days'],
            'trend': self.calculate_trend(week_data['days'])
        }
    
    def generate_monthly_report(self, cost_data: Dict[str, Any], now: datetime) -> Dict[str, Any]:
        """Generate monthly cost report."""
        current_month = now.strftime('%Y-%m')
        month_data = cost_data.get('monthly', {}).get(current_month, {'total': 0.0, 'bots': {}})
        
        # Calculate projection
        day_of_month = now.day
        # Get days in month
        if now.month == 12:
            next_month = now.replace(year=now.year + 1, month=1, day=1)
        else:
            next_month = now.replace(month=now.month + 1, day=1)
        days_in_month = (next_month - timedelta(days=1)).day
        projected_total = (month_data.get('total', 0.0) / day_of_month) * days_in_month if day_of_month > 0 else 0.0
        
        return {
            'month': current_month,
            'total': month_data.get('total', 0.0),
            'projected': projected_total,
            'remaining': 100.0 - month_data.get('total', 0.0),  # Assuming $100 budget
            'bots': month_data.get('bots', {}),
            'top_spenders': self.get_top_spenders(month_data.get('bots', {}), 10),
            'budget_status': self.get_budget_status(month_data.get('total', 0.0), projected_total),
            'recommendations': self.get_recommendations(month_data, projected_total)
        }
    
    def get_top_spenders(self, bots: Dict[str, float], limit: int) -> List[Dict[str, Any]]:
        """Get top spending bots with percentages."""
        total = sum(bots.values())
        
        # Guard against division by zero
        if total == 0:
            return []
        
        sorted_bots = sorted(bots.items(), key=lambda x: x[1], reverse=True)
        return [
            {
                'bot': bot,
                'cost': cost,
                'percentage': f'{(cost / total * 100):.1f}'
            }
            for bot, cost in sorted_bots[:limit]
        ]
    
    def check_daily_alerts(self, daily_data: Dict[str, Any]) -> List[Dict[str, str]]:
        """Check for daily spending alerts."""
        alerts = []
        total = daily_data.get('total', 0.0)
        
        if total > 5:
            alerts.append({
                'level': 'warning',
                'message': f'Daily spend (${total:.2f}) exceeds $5 threshold'
            })
        
        # Check for unusual bot activity
        for bot, cost in daily_data.get('bots', {}).items():
            if cost > 2:
                alerts.append({
                    'level': 'info',
                    'message': f'{bot} spent ${cost:.2f} today (above $2 threshold)'
                })
        
        return alerts
    
    def calculate_trend(self, daily_data: Dict[str, float]) -> str:
        """Calculate spending trend."""
        values = list(daily_data.values())
        if len(values) < 2:
            return 'insufficient_data'
        
        mid = len(values) // 2
        first_half = values[:mid]
        second_half = values[mid:]
        
        first_avg = sum(first_half) / len(first_half) if first_half else 0
        second_avg = sum(second_half) / len(second_half) if second_half else 0
        
        if first_avg == 0:
            return 'insufficient_data'
        
        change = ((second_avg - first_avg) / first_avg) * 100
        
        if change > 20:
            return 'increasing_rapidly'
        elif change > 5:
            return 'increasing'
        elif change < -20:
            return 'decreasing_rapidly'
        elif change < -5:
            return 'decreasing'
        else:
            return 'stable'
    
    def get_budget_status(self, current: float, projected: float) -> Dict[str, Any]:
        """Get budget status information."""
        budget = 100.0  # $100 monthly budget
        percent_used = (current / budget) * 100 if budget > 0 else 0
        projected_percent = (projected / budget) * 100 if budget > 0 else 0
        
        # Calculate days remaining in month
        now = datetime.now()
        if now.month == 12:
            next_month = now.replace(year=now.year + 1, month=1, day=1)
        else:
            next_month = now.replace(month=now.month + 1, day=1)
        days_in_month = (next_month - timedelta(days=1)).day
        days_remaining = days_in_month - now.day
        
        if projected_percent > 100:
            status = 'over_budget'
        elif projected_percent > 90:
            status = 'near_limit'
        elif projected_percent > 75:
            status = 'on_track'
        else:
            status = 'under_budget'
        
        return {
            'percent_used': f'{percent_used:.1f}',
            'projected_percent': f'{projected_percent:.1f}',
            'status': status,
            'days_remaining': days_remaining
        }
    
    def get_recommendations(self, month_data: Dict[str, Any], projected_total: float) -> List[Dict[str, str]]:
        """Get cost optimization recommendations."""
        recommendations = []
        budget = 100.0
        
        if projected_total > budget * 1.1:
            recommendations.append({
                'priority': 'high',
                'action': 'Reduce usage of expensive models',
                'impact': f'Could save ${((projected_total - budget) * 0.5):.2f}'
            })
        
        # Find expensive bots
        bots = month_data.get('bots', {})
        expensive_bots = [
            bot for bot, cost in bots.items()
            if cost > budget * 0.2
        ]
        
        if expensive_bots:
            recommendations.append({
                'priority': 'medium',
                'action': f'Optimize {", ".join(expensive_bots)} - using >20% of budget each',
                'impact': 'Better cost distribution'
            })
        
        if projected_total < budget * 0.5:
            recommendations.append({
                'priority': 'low',
                'action': 'Consider using more capable models for better results',
                'impact': 'Improved quality without exceeding budget'
            })
        
        return recommendations
    
    def format_markdown_report(self, report: Dict[str, Any]) -> str:
        """Format report as Markdown."""
        markdown = f"# 💰 AI Team Cost Report\n\n"
        generated_time = datetime.fromisoformat(report['generated']).strftime('%Y-%m-%d %H:%M:%S')
        markdown += f"**Generated**: {generated_time}\n"
        markdown += f"**Period**: {report['period']}\n"
        markdown += f"**Budget**: ${report['budget']['monthly_limit']}/month\n\n"
        
        data = report['data']
        period = report['period']
        
        if period == 'daily':
            markdown += f"## 📊 Daily Report ({data['date']})\n\n"
            markdown += f"**Total Spend**: ${data['total']:.2f}\n\n"
            
            if data['top_spenders']:
                markdown += "### 🏆 Top Spenders\n\n"
                markdown += "| Bot | Cost | % of Total |\n"
                markdown += "|-----|------|------------|\n"
                for spender in data['top_spenders']:
                    markdown += f"| {spender['bot']} | ${spender['cost']:.2f} | {spender['percentage']}% |\n"
                markdown += "\n"
            
            if data['alerts']:
                markdown += "### ⚠️ Alerts\n\n"
                for alert in data['alerts']:
                    markdown += f"- **{alert['level'].upper()}**: {alert['message']}\n"
        
        elif period == 'weekly':
            markdown += f"## 📊 Weekly Report ({data['start_date']} to {data['end_date']})\n\n"
            markdown += f"**Total Spend**: ${data['total']:.2f}\n"
            markdown += f"**Daily Average**: ${data['daily_average']:.2f}\n"
            markdown += f"**Trend**: {data['trend'].replace('_', ' ')}\n\n"
            
            markdown += "### 📈 Daily Breakdown\n\n"
            markdown += "```\n"
            for date, cost in sorted(data['daily_breakdown'].items()):
                bar = '█' * int(cost * 2)
                markdown += f"{date}: {bar} ${cost:.2f}\n"
            markdown += "```\n\n"
            
            if data['top_spenders']:
                markdown += "### 🏆 Top Spenders\n\n"
                markdown += "| Bot | Cost | % of Total |\n"
                markdown += "|-----|------|------------|\n"
                for spender in data['top_spenders']:
                    markdown += f"| {spender['bot']} | ${spender['cost']:.2f} | {spender['percentage']}% |\n"
        
        elif period == 'monthly':
            markdown += f"## 📊 Monthly Report ({data['month']})\n\n"
            markdown += f"**Current Spend**: ${data['total']:.2f} ({data['budget_status']['percent_used']}% of budget)\n"
            markdown += f"**Projected Total**: ${data['projected']:.2f} ({data['budget_status']['projected_percent']}%)\n"
            markdown += f"**Budget Status**: {data['budget_status']['status'].replace('_', ' ').upper()}\n"
            markdown += f"**Days Remaining**: {data['budget_status']['days_remaining']}\n\n"
            
            # Visual budget bar
            used = int(float(data['budget_status']['percent_used']) / 5)
            projected = int(float(data['budget_status']['projected_percent']) / 5)
            markdown += "### 💵 Budget Visualization\n\n"
            markdown += f"Current:   {'█' * used}{'░' * (20 - used)} {data['budget_status']['percent_used']}%\n"
            markdown += f"Projected: {'█' * projected}{'░' * (20 - projected)} {data['budget_status']['projected_percent']}%\n\n"
            
            if data['top_spenders']:
                markdown += "### 🏆 Top Spending Bots\n\n"
                markdown += "| Bot | Cost | % of Total |\n"
                markdown += "|-----|------|------------|\n"
                for spender in data['top_spenders']:
                    markdown += f"| {spender['bot']} | ${spender['cost']:.2f} | {spender['percentage']}% |\n"
                markdown += "\n"
            
            if data['recommendations']:
                markdown += "### 💡 Recommendations\n\n"
                for rec in data['recommendations']:
                    markdown += f"**{rec['priority'].upper()}**: {rec['action']}\n"
                    markdown += f"- Impact: {rec['impact']}\n\n"
        
        repository = os.environ.get('GITHUB_REPOSITORY', 'owner/repo')
        markdown += "\n---\n"
        markdown += f"*Cost tracking by CFO Bot • [View Full Dashboard](https://github.com/{repository}/actions/workflows/cost-tracker.yml)*"
        
        return markdown


def main():
    """CLI interface."""
    period = sys.argv[1] if len(sys.argv) > 1 else 'daily'
    
    reporter = CostReporter()
    try:
        report = reporter.generate_report(period)
        markdown = reporter.format_markdown_report(report)
        print(markdown)
        sys.exit(0)
    except Exception as error:
        print(f'Error generating report: {error}', file=sys.stderr)
        sys.exit(1)


if __name__ == '__main__':
    main()

